## Doramas Báez Bernal - análisis de opiniones

El objetivo de este trabajo es el análisis de diferentes técnicas de procesamiento de lenguaje natural para un problema de análisis de opiniones (Sentiment Analisys). Para ello, se dispone de un conjunto de datos de opiniones de comida de Amazon con más de 500.000 instancias.

### Objetivos
- Tarea obligatoria: realizar análisis utilizando el proceso NLP clásico considerando el texto original y dos preprocesados y dos características diferentes para representar la matriz término-documento. 
- Tarea optativa: aplicar métodos conexionistas aplicados en NLP

In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
import plotly.express as px


import nltk

from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, recall_score, precision_score

from sklearn.feature_extraction.text import CountVectorizer, TfidfVectorizer

from sklearn.preprocessing import LabelEncoder

from sklearn.ensemble import RandomForestClassifier
from sklearn.svm import SVC

In [2]:
nltk.download('punkt')

[nltk_data] Downloading package punkt to
[nltk_data]     C:\Users\adrian\AppData\Roaming\nltk_data...
[nltk_data]   Package punkt is already up-to-date!


True

In [3]:
df = pd.read_csv('Reviews.csv')
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 568454 entries, 0 to 568453
Data columns (total 10 columns):
 #   Column                  Non-Null Count   Dtype 
---  ------                  --------------   ----- 
 0   Id                      568454 non-null  int64 
 1   ProductId               568454 non-null  object
 2   UserId                  568454 non-null  object
 3   ProfileName             568438 non-null  object
 4   HelpfulnessNumerator    568454 non-null  int64 
 5   HelpfulnessDenominator  568454 non-null  int64 
 6   Score                   568454 non-null  int64 
 7   Time                    568454 non-null  int64 
 8   Summary                 568427 non-null  object
 9   Text                    568454 non-null  object
dtypes: int64(5), object(5)
memory usage: 43.4+ MB


In [4]:
df.isnull().sum()

Id                         0
ProductId                  0
UserId                     0
ProfileName               16
HelpfulnessNumerator       0
HelpfulnessDenominator     0
Score                      0
Time                       0
Summary                   27
Text                       0
dtype: int64

### Dataset

El conjunto de datos con el que se está tratando, está conformado por 568.454 reseñas de alimentos vendidos por amazon. Estas reseñas, incluyen información sobre el tipo de producto, el usuario, la calificación y texto plano contando la experiencia del usuario con el producto.

En principio, no se utilizará el conjunto de datos completo, puesto que procesar 568454 muestras implica un elevado coste computacional. Por lo tanto, para reducir el tamaño del conjunto de datos se realizaran dos pasos:

1. Eliminar los valores nulos
2. Reducir la dimesionalidad del conjunto de datos a unas 100.000 muestras 


In [5]:
df.isnull().sum()
df.dropna(how='any',inplace=True)
print("Dataset sin valores nulos:", df.shape)

df = df[:100000]
print("Dataset reducido a 100000 muestras: ", df.shape)

Dataset sin valores nulos: (568411, 10)
Dataset reducido a 100000 muestras:  (100000, 10)


In [6]:
fig = px.pie(df, names='Score', title='Percentage of the score of the products')
fig.show()

## Revisar

Analizando gráficamente las puntuaciones dadas a los productos, podemos estimar que debería existir un pensamiento positivo antes ellos.

Para realizar el analisis semanticos nos quedaremos con las colmnas 'score' y 'Text' 

In [7]:
print("hola")

hola
